### ISSUE TITLE
Is there a feature to average gradients across batches?

### ISSUE THREAD

**Issue by zzzacwork** [k2-fsa/icefall#426]
I am trying to fine tune the model with pruned_transducer_stateless recipe, but the model diverges after several iterations, 

I am wondering if there is a feature to average gradients across several training batches, and maybe a parameter to set the number of batches to average on.

**Comment by luomingshuang** [comment]
Do you mean the model can accumulate some iterations gradients such as (g1, g2, g3, g4) without parameters update, and after 4 iterations, use the average gradient ( g_avg=(g1+g2+g3+g4)/4 ) to update the parameters?

**Comment by luomingshuang** [comment]
About finetuning for a pretrained pruned_transducer_stateless model, I think maybe you can try let model train more iterations, and then have a look for if the model diverges.

> I am trying to fine tune the model with pruned_transducer_stateless recipe, but the model diverges after several iterations,

**Comment by danpovey** [comment]
@zzzacwork the beta1 parameter in Adam is momentum, that is what is responsible for gradient averaging.
The default is 0.9.  
Using a lower learning rate would probably be a safer bet though, than messing with the momentum parameter.
I think that model can diverge because the parameter tensors get too large, forcing the layernorm scales  to get small to compensate, and when the layernorm scales get too small the learning becomes unstable because it's too sensitive to their values.  Best to not let the parameters get too large; lower learning rate can accomplish that.
Our more recent models are not subject to this particular problem.

**Comment by luomingshuang** [comment]
Oh, yes. I am also doing a fine-tune experiments #413. When I load the pretrained model and decrease the learning rate, the fine-tuning training can converge well.

**Comment by danpovey** [comment]
You should probably start with the final learning rate from the 1st stage of training, or 20% higher, and either leave it constant or very slowly decrease it.

**Comment by zzzacwork** [comment]
> Do you mean the model can accumulate some iterations gradients such as (g1, g2, g3, g4) without parameters update, and after 4 iterations, use the average gradient ( g_avg=(g1+g2+g3+g4)/4 ) to update the parameters?

yes, that's exactly what I meant, huggingface's transformer has something similar, called `args.gradient_accumulation_steps`
https://github.com/huggingface/transformers/blob/0412f3d9298cdb8ba7f69570753ec6a07d240c87/examples/run_squad.py#L221

**Comment by zzzacwork** [comment]
> That is probably just for reproducibility, so that if you have a GPU with less memory you can exactly reproduce an experiment done on a GPU with more memory.

Yeah, I agree, generally speaking, it offers a way to use larger batch size without requiring increasing size of GPU memory accordingly. Larger batch size could possibly make training more stable. Similarly lowering lr rate might work for me at this moment as well I think, I will try it  and see if it would converge on my training data.

**Comment by funboarder13920** [comment]
If you need gradient accumulation I think you can execute `optimizer.step()` and `optimizer.zero_grad()` in an if statement based on a your gradient_accumulation param : https://github.com/k2-fsa/icefall/blob/5379c8e9fa13f6f2364b4a0db89fa3074266fb58/egs/librispeech/ASR/pruned_transducer_stateless/train.py#L651

**Comment by zzzacwork** [comment]
> If you need gradient accumulation I think you can execute `optimizer.step()` and `optimizer.zero_grad()` in an if statement based on a your gradient_accumulation param :
> 
> https://github.com/k2-fsa/icefall/blob/5379c8e9fa13f6f2364b4a0db89fa3074266fb58/egs/librispeech/ASR/pruned_transducer_stateless/train.py#L651

Thanks and I am about to add that, just want check if that is already implemented or there is a better approach for this task.
